sdk/identity/azure-identity/src/private/token_cache_internals.hpp

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+/**
+ * @file
+ * @brief Token cache internals and test hooks.
+ */
+
+#pragma once
+
+#include "token_cache.hpp"
+
+#include <azure/core/credentials/credentials.hpp>
+
+#include <functional>
+#include <map>
+#include <memory>
+#include <shared_mutex>
+#include <string>
+#include <tuple>
+
+namespace Azure { namespace Identity { namespace _detail {
+  /**
+   * @brief Implements internal aspects of token cache and provides test hooks.
+   *
+   */
+  class TokenCache::Internals final {
+    Internals() = delete;
+    ~Internals() = delete;
+
+  public:
+    /**
+     * @brief Represents a unique set of characteristics that are used to distinguish between cache
+     * entries.
+     *
+     */
+    struct CacheKey
+    {
+      std::string TenantId; ///< Tenant ID.
+      std::string ClientId; ///< Client ID.
+      std::string AuthorityHost; ///< Authority Host.
+      std::string Scopes; ///< Authentication Scopes as a single string.
+
+      bool operator<(TokenCache::Internals::CacheKey const& other) const
+      {
+        return std::tie(TenantId, ClientId, AuthorityHost, Scopes)
+            < std::tie(other.TenantId, other.ClientId, other.AuthorityHost, other.Scopes);
+      }
+    };
+
+    /**
+     * @brief Represents immediate cache value (token) and a synchronization primitive to handle its
+     * updates.
+     *
+     */
+    struct CacheValue
+    {
+      std::shared_timed_mutex ElementMutex;
```

Comment:
I don't see why we have to use **timed** mutex rather than just `std::shared_mutex`.

Reply:
`std::shared_timed_mutex` is C++14, while `std::shared_mutex` is C++17. We can't go newer than 14.

Reply:
ok